internal/constants/scripts.go

```diff
@@ -196,3 +195,27 @@ if [ "$host_dir" ]
 fi
 echo "PreNetworkConfig End"
 `
+
+const PreNetworkConfigService = `
+[Unit]
+Description=Assisted static network config
+DefaultDependencies=no
+
+# We need to run after the kernel and udev have detected all the network interface devices,
+# otherwise the mechanism that we use to find the MAC addresses will not work and the configuration
+# will not be applied.
+After=systemd-udev-settle.service
+
+# We need to run after this service because it runs the 'nm-initrd-generator' tool, which removes
+# all the connections and re-creates them from kernel command line arguments.
+After=coreos-enable-network.service
+
+# We need to run before this service so that when it starts all the connection files that we want
+# have already been created and they don't need to be re-loaded.
```

Comment:
If we expect static configuration to be injected, perhaps what's actually simplest is to inject configuration that disables DHCP first?

https://docs.fedoraproject.org/en-US/fedora-coreos/sysconfig-network-configuration/#_disabling_automatic_configuration_of_ethernet_devices

Then we know NM will start up as usual, but *do nothing* until explicitly told to by this service.

And at that point, this service can wait for the devices it expects.

Reply:
I don't fully understand everything that's going on in the flow here, but basically I think a flow of having a separate systemd unit run that's ordered *after* `NetworkManager.service` and injects config and then runs `nmcli reload` is actually better; the key is to just disable DHCP by default.

Reply:
I think that "no-auto-default" configuration is already there by default:

```
# cat usr/lib/NetworkManager/conf.d/initrd-no-auto-default.conf 
[.config]
enable=env:initrd

[main]
no-auto-default=*
```

But the "coreos-enable-network" service brings the "default_connection" back.

Reply:
> But the "coreos-enable-network" service brings the "default_connection" back.

Oooh.  That seems like a bug.  Or at least, something we need to enable suppressing.

Reply:
is `usr/lib/NetworkManager/conf.d/initrd-no-auto-default.conf` in the initramfs?

Reply:
> is `usr/lib/NetworkManager/conf.d/initrd-no-auto-default.conf` in the initramfs?

Yes:

```
# mount -o loop,ro /files/images/rhcos-4.13.0-rc.2-x86_64-live.x86_64.iso /mnt
# lsinitrd /mnt/images/pxeboot/initrd.img | grep no-auto
-rw-r--r--   1 root     root           54 Jan  1  1970 usr/lib/NetworkManager/conf.d/initrd-no-auto-default.conf
```

Reply:
I can't really find where that's defined in this repo: https://github.com/openshift/assisted-service/search?q=no-auto-default

Reply:
oh actually I think it comes from NetworkManager itself: https://github.com/dracutdevs/dracut/blob/master/modules.d/35network-manager/initrd-no-auto-default.conf

ok, yes. This is by design. NM relies on `nm-initrd-generator` to generate the connections in the initramfs.

Reply:
see my suggestion in https://github.com/openshift/assisted-service/pull/5107#issuecomment-1498005927

Reply:
> see my suggestion in [#5107 (comment)](https://github.com/openshift/assisted-service/pull/5107#issuecomment-1498005927)

I will try that, thanks @dustymabe.

Reply:
@dustymabe I quickly tried your suggestion, and unfortunately it doesn't work. The reason is that we read the `/sys/class/net/*` directories to find the MAC addresses of the network interfaces, and at the point that `coreos-copy-firstboot-network.service` runs those directories aren't populated yet. We need to know the MAC addresses of the devices because we generate a single ISO for all the hosts, containing the configuration files for all of them, and then we use the MAC addresses to determine what files should be applied in each particular host.

Reply:
> and at the point that `coreos-copy-firstboot-network.service` runs those directories aren't populated yet

Could you add dependencies to your service such that those things exist but then also still ues `Before=coreos-copy-firstboot-network.service` or does that create an ordering loop?

Reply:
If I add `After=systemd-udev-settle.service` things seem to work correctly @dustymabe, so I assume there are no loops. I will update the pull request with that change.

Reply:
Cool. Glad it's working for you!